src/codec.rs

```diff
@@ -554,25 +552,42 @@ impl<T> WrapperTypeDecode for Box<T> {
 		//
 		//         Constructing a `Box` from `NonNull::dangling` is also always safe as long
 		//         as the underlying type is zero-sized.
-		let mut boxed: Box<MaybeUninit<T>> = unsafe { Box::from_raw(ptr) };
-
+		let mut boxed = unsafe { Box::from_raw(ptr) };
 		T::decode_into(input, &mut boxed)?;
 
 		// Decoding succeeded, so let's get rid of `MaybeUninit`.
 		//
 		// TODO: Use `Box::assume_init` once that's stable.
-		let ptr: *mut MaybeUninit<T> = Box::into_raw(boxed);
-		let ptr: *mut T = ptr.cast();
+		let ptr: *mut T = Box::into_raw(boxed).cast();
 
 		// SAFETY: `MaybeUninit` doesn't affect the memory layout, so casting the pointer back
 		//         into a `Box` is safe.
-		let boxed: Box<T> = unsafe { Box::from_raw(ptr) };
+		let boxed = unsafe { Box::from_raw(ptr) };
```

Comment:
Please don't delete the explicit types nor "simplify" this code; they're here for readability in what is a highly tricky piece of code, *and* to make sure the compiler statically checks things for us.

We do *not* want the compiler to automatically infer things here. It might still work after being simplified, but then the next person will come, make a change, and then subtly break this code. The seemingly redundant types are a safeguard against that happening.

Reply:
restored